### PR TITLE
Skip files when they are more than 15000 lines

### PR DESCRIPTION
**Pull request by ariskataoka**
TLDR; Some builds are currently timing out due to PRs that contain too large files. 

This PR addresses the [issue 147](https://github.com/Automattic/vip-go-ci/issues/147).

If the PR contains files with more than 15000 lines, the lint scan and the phpcs will skip it. 

- Adds logic to validate file size;
- Adds logic to modify the bot message;
- Adds logic to ignore files that have been skipped previously;

TODO:
- [ X ] Implement [new feature / logic]
- [ X ] Add/update unit-tests
- [ X ] Manual testing
 - [ X ] Pull-Request with files with more than 15000 lines

Obs.: The build testing has constantly been failing due to GH API limit rate.

```
[ 2021-05-26T08:37:48+00:00 -- 0 ]  Ran out of request limits for GitHub, cannot continue without making further requests.; {
    "github_url": "https:\/\/api.github.com\/repos\/automattic\/vip-go-ci-testing\/pulls\/39\/commits?page=1&per_page=100",
    "x-ratelimit-remaining": "1",
    "x-ratelimit-limit": "60"
}
```

### CODE REVIEW

File: defines.php [hunk 1]
```diff
@@ -140,7 +140,23 @@
...
 define( 'VIPGOCI_GIT_DIFF_DATA_SOURCE_GITHUB_API',	'github-api' );
 
+
+
+/**
+ * Define file number of lines limit
+ */
+define( 'VIPGOCI_SKIPPED_FILES',	        'skipped-files'	);
+
+define( 'VIPGOCI_VALIDATION_MAXIMUM_LINES_LIMIT',   15000   );
```
Comment (gudmdharalds) [inline defines.php:154]: Maybe use this as a default, but then be able to customise the number on the command-line or via environmental options? I do see us possibly changing this number, would be good not having to change the code to do that.

File: file-validation.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	 * if it is > than the default limit (defined at defined)
+	 *
+	 * the bot won't scan it
+	 */
+	$cmd = sprintf(
+		'wc -l %s | awk \'{print $1;}\'',
+		escapeshellcmd( $temp_file_name )
+	);
+
+	vipgoci_log( 'Validating file number of lines before bot runs', array( 'cmd' => $cmd ), 0 );
```
Comment (gudmdharalds) [inline file-validation.php:51]: Suggest to skip the `before bot runs` as processing has already started at this point.

File: file-validation.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+	$output = ( int ) vipgoci_runtime_measure_shell_exec(
+		$cmd,
+		'file_validation'
+	);
+
+	vipgoci_log( 'Validation file number of lines output', array( 'output' => $output ), 0 );
```
Comment (gudmdharalds) [inline file-validation.php:58]: Suggest to include file-name as well in the logging, slightly easier to see what file it is.

File: main.php [hunk 4]
```diff
@@ -41,7 +41,14 @@ function vipgoci_exit_status( $results ) {
 				return VIPGOCI_EXIT_CODE_ISSUES;
 			}
 		}
+	}
 
+	if ( ! empty( $results['skipped-files'] ) ) {
+		foreach ( $results['skipped-files'] as $pr_number ) {
+			if( 0 < $pr_number[ 'total' ] ) {
+				return VIPGOCI_EXIT_CODE_ISSUES;
```
Comment (gudmdharalds) [inline main.php:49]: Suggest to explain in one sentence why the program returns non-zero here.

**Review by gudmdharalds — CHANGES_REQUESTED** [review]
Thank you for the Pull-Request! This looks very good, just left a few a few comments.

File: file-validation.php (hunk 3 above)
Comment (ariskataoka) [reply file-validation.php:58]: That makes sense. I will add it.

**Review by ariskataoka — COMMENTED** [review]
Thanks for reviewing. I'm addressing the points

File: phpcs-scan.php [hunk 5]
```diff
@@ -212,6 +212,24 @@ function vipgoci_phpcs_scan_single_file(
...
+	 */
+	if ( 0 !== $validation[ 'total' ] ) {
+		$skipped = array(
+			'file_issues_arr_master'	=> array(),
+			'file_issues_str'		=> array(),
+			'temp_file_name'		=> $temp_file_name,
+			'validation'            => $validation
+		);
+		unlink( $temp_file_name );
+
```
Comment (gudmdharalds) [inline phpcs-scan.php:228]: In `lint-scan.php` [above](https://github.com/Automattic/vip-go-ci/pull/176/files#diff-8a7186458f366d2d94ecd363f6f23db1b629ffba464ff2821fa3961f7cd57b9fR291) the `vipgoci_set_prs_implicated_skipped_files()` function is called, but not here.

We want the scanning modules (lint, phpcs, svg) to be independent of each other, and end-users being able to enable and disable them as they wish. If linting is disabled, it seems this function is never called. Possibly this function needs to be called here too?

File: file-validation.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	 * if it is > than the default limit (defined at defined)
+	 *
+	 * the bot won't scan it
+	 */
+	$cmd = sprintf(
+		'wc -l %s | awk \'{print $1;}\'',
+		escapeshellcmd( $temp_file_name )
+	);
+
+	$output = ( int ) vipgoci_runtime_measure_shell_exec(
```
Comment (gudmdharalds) [inline file-validation.php:51]: Maybe check if `$output` is a numeric before casting, and if it is not print error? Otherwise error messages might lead to strange problems later on as casting a string can result in a non-zero integer value.

**Review by gudmdharalds — CHANGES_REQUESTED** [review]
Hi!

Thanks again for the update. This is great. I left a few comments on small things.

File: file-validation.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ * @return array
+ *	[
+ *		'issues' =>
+ *		[
+ *			'max-lines' => [$file_name],
+ *		],
+ *		'total' => 1
+ *	]
+ */
+function vipgoci_validate( $temp_file_name, $file_name ) {
```
[the remark this answers is not in the document]
Comment (ariskataoka) [reply file-validation.php:21]: I'm more than happy to address this - I'm a "strict type person". I didn't add because I saw most of the functions in this project don't have strict types and wanted to keep consistency).

I also think we can enforce strict_types for the new files/classes at the beginning.

File: file-validation.php (hunk 6 above)
Comment (ariskataoka) [reply file-validation.php]: I made a small modification to avoid that casting. Can you take a look and let me know what your thoughts are?

File: file-validation.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+function vipgoci_is_number_of_lines_valid( $temp_file_name, $file_name ) {
+	/**
+	 * Calculates the file number of lines
+	 *
+	 * if it is > than the default limit (defined at defined)
+	 *
+	 * the bot won't scan it
+	 */
+	$cmd = sprintf(
+		'wc -l %s | awk \'{print $1;}\'',
```
[the remark this answers is not in the document]
Comment (ariskataoka) [reply file-validation.php]: Yes, that makes sense.  THanks for sharing that issue here.

File: file-validation.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+/**
+ * @param $temp_file_name
+ * Validates if the number of lines is lower than the maximun
+ *
+ * @return bool
+ */
+function vipgoci_is_number_of_lines_valid( $temp_file_name, $file_name ) {
+	/**
```
[the remark this answers is not in the document]
Comment (ariskataoka) [reply file-validation.php:43]: That's a great suggestion. I didn't realize that this project contains a cache layer. Gonna implement that.

File: file-validation.php (hunk 6 above)
Comment (gudmdharalds) [reply file-validation.php]: Looks good! :)

File: phpcs-scan.php (hunk 5 above)
Comment (ariskataoka) [reply phpcs-scan.php:228]: That would make sense. However, at this stage, the logic for phpcs-scan is slightly different than the lint-scan. While the calls to create the temp file are done in the prs_implicated iteration for phpcs-scan, the lint one is done per commits; adding that function call would add a few more unnecessary array iterations (for now). 
Also, the project is procedural and, even because of the sequence of execution, even when the lint is disabled, the php-cs will validate the files. So, I wouldn't concern too much about it now.

**Review by gudmdharalds — CHANGES_REQUESTED** [review]
This looks very good. 

I left a few comments regarding the caching logic which I think can be simplified. Also left comments on usage of  `assertEquals()` in the unit-tests.

Once the above is improved, and being able to configure maximum number of lines via the command-line, I think this is ready. 

We will probably have to look into the unit-tests and see if we can try to reduce the number of GitHub API calls -- I suspect we are reaching a limit with each test run.

File: file-validation.php [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+		vipgoci_runtime_measure_shell_exec( $cmd, 'file_validation' )
+	);
+
+	vipgoci_log(
+		sprintf( 'Validation number of lines ', $file_name ),
+		array( 'file_name' => $file_name, 'cmd' => $cmd, 'output' => $output ), 0
+	);
+
+	$is_number_of_lines_valid = vipgoci_verify_number_of_lines_output( $output );
+	vipgoci_cache_set_is_number_of_lines_valid_by_file_name( $cache_key, $current_cached_value, $file_name, $is_number_of_lines_valid );
```
[the remark this answers is not in the document]
Comment (ariskataoka) [reply file-validation.php:76]: That would turn the ``commit_key`` to be something like ``someprefix-$commit_id-$filename`` , which is also not ideal, but would do the job.

File: file-validation.php [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+	return $is_number_of_lines_valid;
+}
+
+/**
+ * @param string $output
+ *
+ * @return bool
+ */
+function vipgoci_verify_number_of_lines_output( string $output ): bool
```
[the remark this answers is not in the document]
Comment (ariskataoka) [reply file-validation.php:86]: I wanted to make sure that the functions I'm implementing are doing only one single thing.
It's not a good practice to create long functions; to make this file "tell a history" to the reader, I'm trying to keep them small, making it easier to understand.

[the remark this answers is not in the document]
Comment (gudmdharalds) [reply file-validation.php:86]: I understand, this is absolutely fine. Just wanted to suggest a different solution. :slightly_smiling_face:

File: file-validation.php [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+	 * Verifies if number of lines validation are in the cache
+	 * If so, returns the value
+	 */
+	$cache_key = vipgoci_cache_get_is_number_of_lines_valid_key( $commit_id );
+	$current_cached_value = vipgoci_cache_get_is_number_of_lines_valid( $cache_key );
+	$is_number_of_lines_valid = vipgoci_cache_get_is_number_of_lines_valid_by_file_name( $current_cached_value, $file_name );
+	if ( ! is_null( $is_number_of_lines_valid ) ) {
+		return $is_number_of_lines_valid;
+	}
+
```
[the remark this answers is not in the document]
Comment (ariskataoka) [reply file-validation.php:57]: Just to check if I understand. Are you suggesting the implementation of an array as a ``cache_key`` ?

[the remark this answers is not in the document]
Comment (gudmdharalds) [reply file-validation.php:57]: That is correct, using array is supported. It is a bit unusual. :slightly_smiling_face: 

However, it is possible to use a string as well. In fact, the array will be converted to a string by `vipgoci_cache()`.

[the remark this answers is not in the document]
Comment (ariskataoka) [reply file-validation.php:57]: Yes, after debugging it, I could see it's converted to a son value. Thanks for clarifying!

File: phpcs-scan.php (hunk 5 above)
Comment (ariskataoka) [reply phpcs-scan.php:228]: Rewriting my previous comment in other words, what I meant is:

[lint-scan.php calls `vipgoci_set_prs_implicated_skipped_files`](https://github.com/Automattic/vip-go-ci/pull/176/files#diff-8a7186458f366d2d94ecd363f6f23db1b629ffba464ff2821fa3961f7cd57b9fR292)  to set all the implicated files at the end because at that point, the application knows that value; 

[phpcs-scan.php  calls ``vipgoci_set_skipped_file``](https://github.com/Automattic/vip-go-ci/pull/176/files#diff-3c410336e93f47dad45b2700845c92749e35de8c1a8d82b8e8f1d312100ca4e8R590) instead. Because of the way it iterates over the ``$prs_implicated`` (pre-existent logic); 

If we observe the ``vipgoci_set_prs_implicated_skipped_files`` function closely, we can see that it [calls the same  ``vipgoci_set_skipped_file`` function](https://github.com/Automattic/vip-go-ci/pull/176/files#diff-d324959e718f3d5d8ac49fed3363985d839138080b4456e85a82bd2181c7400cR47) that ``phpcs-scan`` calls. 

Having that said, both would work whether one or another is disabled. Do you think that is not enough?

Also, by ``verify that the skipping feature works for PHPCS when PHP Linting is disabled (and vice versa)``, do you mean creating unit tests?

Comment (gudmdharalds) [reply phpcs-scan.php:228]: > Having that said, both would work whether one or another is disabled. Do you think that is not enough?

My apologies, I misread the logic before. There is nothing that needs to be done here. 

> Also, by verify that the skipping feature works for PHPCS when PHP Linting is disabled (and vice versa), do you mean creating unit tests?

I meant by running `vip-go-ci` with these different combinations of options, just manual testing.
